app/services/ui_notifications/report_finished.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module UINotifications
+  class ReportFinished < UINotifications::Base
+    include Rails.application.routes.url_helpers
+
+    def initialize(composer, job_id)
+      @composer = composer
+      @job_id = job_id
+      super(User.current)
+    end
+
+    private
+
+    def create
+      add_notification
+    end
+
+    def update_notifications
+      blueprint.notifications.
+          where(subject: subject).
+          update_all(expired_at: blueprint.expired_at)
+    end
+
+    def add_notification
+      Notification.create!(
+        initiator: initiator,
+        audience: ::Notification::AUDIENCE_USER,
+        subject: subject,
+        message: N_('Report "%s" is ready to download') % @composer.template.name,
+        notification_blueprint: blueprint,
+        links: [
+          {
+            title: N_('Download Report'),
+            href: report_data_report_template_path(subject, job_id: @job_id),
+          },
+          {
+            title: N_('Regenerate Report'),
+            href: generate_report_template_path(@composer.template, {report_template_report: { input_values: @composer.to_params[:input_values]}}),
```

Comment:
since you'll need to touch the code anyway, it would be good to be consistent about the white spaces on this line, once you have it after `{` and once not. Also if you decide to add white spaces inside of `{`, the do the same before closing bracket.